### PR TITLE
networkctl: show stats

### PR DESCRIPTION
**Pull request by yuwata**
This implements several features requested in #12657.

### CODE REVIEW

**Comment by yuwata** [comment]
Rebased. PTAL.

File: src/network/networkctl.c [hunk 1]
```diff
@@ -1446,7 +1500,7 @@ static int parse_argv(int argc, char *argv[]) {
         assert(argc >= 0);
         assert(argv);
 
-        while ((c = getopt_long(argc, argv, "ha", options, NULL)) >= 0) {
+        while ((c = getopt_long(argc, argv, "has", options, NULL)) >= 0) {
```
Comment (keszybz) [inline src/network/networkctl.c:1623]: Please add to --help too.

File: src/network/networkctl.c [hunk 2]
```diff
@@ -822,7 +822,7 @@ static int dump_list(Table *table, const char *prefix, char **l) {
         r = table_add_cell(table, NULL, TABLE_EMPTY, NULL);             \
         if (r < 0)                                                      \
                 return r;                                               \
-        r = table_add_cell_full(table, NULL, TABLE_STRING, name,        \
+        r = table_add_cell_full(table, NULL, TABLE_STRING, name ":",    \
```
Comment (keszybz) [inline src/network/networkctl.c]: Misplaced chunk?

File: shell-completion/bash/networkctl [hunk 3]
```diff
@@ -32,7 +32,7 @@ _networkctl() {
     local i verb comps
     local cur=${COMP_WORDS[COMP_CWORD]} prev=${COMP_WORDS[COMP_CWORD-1]}
     local -A OPTS=(
-        [STANDALONE]='-a --all -h --help --version --no-pager --no-legend'
+        [STANDALONE]='-a --all -h --help --version --no-pager --no-legend -s --stats'
```
Comment (keszybz) [inline shell-completion/bash/networkctl:35]: Would you mind doing zsh too?

File: src/network/networkd-link-bus.c [hunk 4]
```diff
@@ -10,11 +10,42 @@
...
+        assert(bus);
+        assert(reply);
+        assert(userdata);
+
+        manager = link->manager;
+
+        if (!manager->use_speed_meter ||          /* disabled */
+            manager->speed_meter_usec_old == 0 || /* not started yet */
+            !link->stats_updated)                 /* error in getting link stats */
+                return sd_bus_message_append(reply, "(dd)", -1, -1);
```
Comment (keszybz) [inline src/network/networkd-link-bus.c]: Shouldn't this be -1., -1. ?

File: src/network/networkd-link-bus.c [hunk 5]
```diff
@@ -10,11 +10,42 @@
...
+
+        return sd_bus_message_append(reply, "(dd)",
+                                     (double) (link->stats_new.tx_bytes - link->stats_old.tx_bytes) * USEC_PER_SEC /
+                                     (double) (manager->speed_meter_usec_new - manager->speed_meter_usec_old),
```
Comment (keszybz) [inline src/network/networkd-link-bus.c]: This second cast is not necessary: double * long / long = double.

File: src/network/networkd-speed-meter.h [hunk 6]
```diff
@@ -0,0 +1,9 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+#pragma once
+
+#define SPEED_METER_DEFAULT_TIME_INTERVAL (10 * USEC_PER_SEC)
+#define SPEED_METER_MINIMUM_TIME_INTERVAL (100 * USEC_PER_MSEC)
```
Comment (keszybz) [inline src/network/networkd-speed-meter.h:8]: Maybe add a comment that those have differnet units on purpose (s vs ms)...

File: src/network/networkd-manager.c [hunk 7]
```diff
@@ -1369,10 +1370,15 @@ int manager_new(Manager **ret) {
         _cleanup_(manager_freep) Manager *m = NULL;
         int r;
 
-        m = new0(Manager, 1);
+        m = new(Manager, 1);
         if (!m)
                 return -ENOMEM;
 
+        *m = (Manager) {
+                .use_speed_meter = true,
```
Comment (keszybz) [inline src/network/networkd-manager.c]: Hmm, I'm not convinced this should be on by default. Waking up every 10 s do a non-trivial operation doesn't seem nice for battery life. Would it be possible to keep this disabled by default and only enable when someone starts listening for the statistics?

File: src/network/networkctl.c [hunk 8]
```diff
@@ -854,6 +909,49 @@ static int dump_statistics(Table *table, const LinkInfo *info) {
         return 0;
 }
 
+static int get_prefix(double val, double *ret_div, char **ret_prefix) {
+        double m = (double) 1000*1000*1000*1000*1000;
```
Comment (keszybz) [inline src/network/networkctl.c]: `double m = 1e15;`?

File: src/network/networkctl.c [hunk 9]
```diff
@@ -854,6 +909,49 @@ static int dump_statistics(Table *table, const LinkInfo *info) {
...
+        m /= 1000;
+        if (val > m) {
+                s = strdup("M");
+                goto finalize;
+        }
+        m /= 1000;
+        if (val > m) {
+                s = strdup("k");
+                goto finalize;
+        }
```
Comment (keszybz) [inline src/network/networkctl.c]: This could be turned into a loop easily...

File: src/network/networkd-manager.c (hunk 7 above)
Comment (keszybz) [reply src/network/networkd-manager.c]: Yeah, this is the biggest issue. I don't think we want yet-another thing waking up in the background constantly. I'd very much prefer to have this opt-in and on-demand.

File: src/network/networkd-link-bus.c (hunk 4 above)
Comment (yuwata) [reply src/network/networkd-link-bus.c]: ~~What do you mean? Should we return some error, e.g. `-EOPNOTSUPP`?~~

OK.

File: shell-completion/bash/networkctl (hunk 3 above)
Comment (yuwata) [reply shell-completion/bash/networkctl:35]: Sorry, I've never used zsh, and am not familiar with zsh-completion...

**Comment by yuwata** [comment]
@keszybz Thank you for your review. I've force-pushed an updated version. I hope all your comments are addressed. PTAL.

File: shell-completion/bash/networkctl (hunk 3 above)
Comment (keszybz) [reply shell-completion/bash/networkctl:35]: OK.

File: src/network/networkd-link-bus.c [hunk 10]
```diff
@@ -10,11 +11,56 @@
...
+        manager = link->manager;
+
+        if (!manager->use_speed_meter)
+                return sd_bus_error_set(error, BUS_ERROR_SPEED_METER_INACTIVE, "Speed meter is disabled.");
+
+        if (manager->speed_meter_usec_old == 0)
+                return sd_bus_error_set(error, BUS_ERROR_SPEED_METER_INACTIVE, "Speed meter is not active.");
+
+        if (!link->stats_updated)
+                return sd_bus_error_set(error, BUS_ERROR_SPEED_METER_INACTIVE, "Failed to measure bit-rates.");
```
Comment (keszybz) [inline src/network/networkd-link-bus.c:40]: "bit rates"?

**Pull request merged**